File: FStarFSharp/CPEG/TypeInference.fs
module TypeInference

open Expressions
open RET

let randomStr = 
    let chars = "ABCDEFGHIJKLMNOPQRSTUVWUXYZ0123456789"
    let charsLen = chars.Length
    let random = System.Random()
    fun len -> 
        let randomChars = [|for i in 0..len -> chars.[random.Next(charsLen)]|]
        new System.String(randomChars)

let freshName (chi : Set<string>)= let candidate = randomStr 2 in if chi.Contains candidate then candidate else candidate+(randomStr 2)


type TypingContext = {
    mutable chi : Set<string>;
    mutable gamma : Map<CPEGExp,RET>;
}
let addChi (ctx:TypingContext) = fun str -> ctx.chi <- Set.add str ctx.chi
let addGamma (ctx:TypingContext) = fun (e,shape) -> ctx.gamma <- Map.add e shape ctx.gamma

let join (p:Map<'a,'b>) (q:Map<'a,'b>) = 
    Map(Seq.concat [ (Map.toSeq p) ; (Map.toSeq q) ])

(*
    getShape ::
    TypingContext := {gamma, chi}
    CPEGExpression := expression to be typed
    RET := Type for CPEGExpression
    MAP(String to RET) := global set of RET
*)

let rec getShape (ctx:TypingContext) = function
    | Empty -> RETEmpty, Map.empty
    | Terminal _ -> RETEmpty, Map.empty
    | Nonterm pg -> match Map.tryFind (Nonterm pg) ctx.gamma with
                    | Some X -> (X, Map.empty)
                    | None -> let Xa = (freshName ctx.chi)
                              addChi ctx Xa
                              addGamma ctx (Nonterm pg, RETVar Xa)
                              getShape ctx pg |> fun (t, globSet) -> (RETVar Xa, Map.add Xa t globSet)
    | Seq exs -> List.foldBack (fun e T -> (getShape ctx e) :: T) exs List.empty
                    |> List.unzip |> fun (a, bx) -> (RETConcat a, List.fold join Map.empty bx)
    | Alt (e1,e2) -> let (shape1, E1) = getShape ctx e1
                     let (shape2, E2) = getShape ctx e2
                     match (shape1, shape2) with
                        | (RETErr, _) -> (RETErr, E1)
                        | (_, RETErr) -> (RETErr, E2)
                        | (_, _) -> (RETUnion (shape1, shape2), join E1 E2)
    | Rep e -> let (shape0,E0) = getShape ctx e
               match shape0 with
                    | RETEmpty -> (shape0, E0)
                    | RETErr -> (RETErr, E0)
                    | derivedRET -> (RETRep derivedRET, E0)
    | Not _ -> (RETEmpty, Map.empty)
    | Cap (l, e) -> let (shape, E0) = getShape ctx e in (RETLabel (l, shape), E0)
    | FCap (e1, l, e2) -> let (shape1, E1) = getShape ctx e1
                          let (shape2, E2) = getShape ctx e2
                          let Xa = freshName ctx.chi
                          addChi ctx Xa
                          RETVar Xa, (join E1 E2 |> Map.add Xa (RETUnion (RETLabel(l, (RETConcat [RETVar Xa; shape2])),shape1)))

File: FStarFSharp/CPEG/TypeInference.py
from Expressions import *
from RET import *
import random
import functools

def randomStr(l):
    chars = "ABCDEFGHIJKLMNOPQRSTUVWUXYZ0123456789"
    charsLen = len(chars)

    randomChars = (chars[random.randrange(charsLen)] for i in range(l))
    return "".join(randomChars)

def freshName (chi, str):
    candidate = str + randomStr(2)
    if candidate in chi:
        return freshName(chi, candidate)
    else:
        chi.add(candidate)
        return candidate

class TypingContext:
    def __init__(self, chi, gamma):
        self.chi = chi
        self.gamma = gamma

def unzip(li):
    s = []
    t = []
    for (i, j) in li:
        s.append(i)
        t.append(j)

    return (s, t)

def addChi(ctx: TypingContext, str):
    ctx.chi.add(str)
def addGamma(ctx: TypingContext, e, shape):
    ctx.gamma[e] = shape
def join(p, q):
    p.update(q)
    return p

#    getShape ::
#    TypingContext := {gamma, chi}
#    CPEGExpression := expression to be typed
#    RET := Type for CPEGExpression
#    MAP(String to RET) := global set of RET

def getShape(ctx: TypingContext, arg: CPEGExp):
    if isinstance(arg, Empty) or isinstance(arg, Terminal) or isinstance(arg, Not):
        return (RETEmpty(), {})
    elif isinstance(arg, Nonterm):
        if arg in ctx.gamma:
            return (X, {})
        else:
            Xa = freshName(ctx.chi, "")
            addChi(ctx, Xa)
            addGamma(ctx, arg, RETVar(Xa))
            (t, globSet) = getShape(ctx, pg)
            globSet[Xa] = t
            return (RETVar(Xa), grobSet)
    elif isinstance(arg, Seq):
        (a, bx) = unzip(functools.reduce(lambda T, e: [getShape(ctx, e)] + T, reversed(arg.elist), []))
        return (RETConcat(a), functools.reduce(lambda p, q: join(p, q), bx, {}))
    elif isinstance(arg, Alt):
        (shape1, E1) = getShape(ctx, arg.e1)
        (shape2, E2) = getShape(ctx, arg.e2)
        if isinstance(shape1, RETErr):
            return (RETErr(), E1)
        elif isinstance(shape2, RETErr):
            return (RETErr(), E2)
        else:
            return (RETUnion(shape1, shape2), join(E1, E2))
    elif isinstance(arg, Rep):
        (shape0, E0) = getShape(ctx, arg.value)
        if isinstance(shape0, RETEmpty) or isinstance(shape0, RETErr):
            return (shape0, E0)
        else:
            return (RETRep(shape0), E0)
    elif isinstance(arg, Cap):
        (shape, E0) = getShape(ctx, arg.e)
        return (RETLabel(arg.l, shape), E0)
    elif isinstance(arg, FCap):
        (shape1, E1) = getShape(ctx, arg.e1)
        (shape2, E2) = getShape(ctx, arg.e2)
        Xa = freshName(ctx.chi, "")
        addChi(ctx, Xa)
        newE = join(E1, E2)
        newE[Xa] = RETUnion(RETLabel(arg.l, (RETConcat([RETVar(Xa), shape2]))), shape1)
        return (RETVar(Xa), newE)
    elif isinstance(arg, Leaf):
        return (RETLeaf(arg.l), {})
        #Type checking for sigma is altavatively executed by F# type system
    else:
        print(type(arg), "ERROR")
        return (RETEmpty(), {})
